### PR TITLE
Bug 1188299 - Sync immediately a tab is sent

### PR DESCRIPTION
**Pull request by fluffyemily**
https://bugzilla.mozilla.org/show_bug.cgi?id=1188299

Also, ensure that commands are sent as JSON and not String

### CODE REVIEW

File: Providers/Profile.swift [hunk 1]
```diff
@@ -296,7 +296,7 @@ public class BrowserProfile: Profile {
         let commands = items.map { item in
             SyncCommand.fromShareItem(item, withAction: "displayURI")
         }
-        self.remoteClientsAndTabs.insertCommands(commands, forClients: clients)
+        self.remoteClientsAndTabs.insertCommands(commands, forClients: clients) >>> { self.syncManager.syncClients() }
```
Comment (rnewman) [inline Providers/Profile.swift]: Do we need to explicitly denote a background operation here (e.g., if we're called from an extension)?

Please also test this with a queued tab ready to receive. I believe the stub delegate (that we'll use if this is called from inside an extension) just throws away the incoming `displayURI` command, and the received command will never be received.

This is the first time we'll be triggering a sync from an extension, so there are some new things to address…

Comment (fluffyemily) [inline Providers/Profile.swift]: Do you mean here that if, when we sync, there are commands queued to execute on _our_ client then those commands will not be executed?

Comment (rnewman) [inline Providers/Profile.swift]: Yes. 

```
    private func getSyncDelegate() -> SyncDelegate {
        if let app = self.app {
            return BrowserProfileSyncDelegate(app: app)
        }
        return CommandDiscardingSyncDelegate()
    }
```

There are various approaches to fixing this.

* Replace `CommandDiscardingSyncDelegate` with one that stuffs incoming tabs into the “View Later” queue (which didn't exist when it was written). This works around the need to have access to the `UIApplication` in order to process incoming commands, and partly eliminates the distinction between in-extension and in-app syncs.
* Don't advance client timestamps when syncing clients from an extension. This is kinda hacky. We'll re-process our own record.
* Expose a way to do an upload-only clients sync. "Upload-only" is a bit of a misnomer, granted.
* Don't trigger an outgoing sync if we're not sending from inside Firefox. That defeats some of the point of this PR.

Comment (fluffyemily) [inline Providers/Profile.swift]: Interestingly, this is already happening.

When we invoke the extension, the first thing it does is perform a client sync to get the latest clients to choose from. We lose any sent commands at this point. Even if we weren't syncing immediately on send tab we'd _still_ be having this problem. Good catch :smile_cat:

**Comment by fluffyemily** [comment]
So, I've gone for the simplest solution first approach:
1. add received commands to tab queue when received during sync in extension
2. ensure that if extension used within the same app (i.e. within firefox & therefore `applicationDidBecomeActive` will not get called for a while) that the tab queue is executed when the extension is closed.

**Pull request merged**